fish-rust/src/env/environment_impl.rs

```diff
@@ -0,0 +1,1225 @@
+        return mismatch;
+    }
+
+    /// Get the exported variables into a variable table.
+    fn get_exported(n: &EnvNodeRef, table: &mut VarTable) {
+        let n = n.borrow();
+
+        // Allow parent scopes to populate first, since we may want to overwrite those results.
+        if let Some(next) = n.next.as_ref() {
+            Self::get_exported(next, table);
+        }
+
+        for (key, var) in n.env.iter() {
+            if var.exports() {
+                // Export the variable. Note this overwrites existing values from previous scopes.
+                table.insert(key.clone(), var.clone());
+            } else {
+                // We need to erase from the map if we are not exporting, since a lower scope may have
+                // exported. See #2132.
+                table.remove(key);
+            }
+        }
+    }
+
+    /// Return a newly allocated export array.
+    fn create_export_array(&self) -> Arc<OwningNullTerminatedArray> {
+        FLOG!(env_export, "create_export_array() recalc");
+        let mut vals = VarTable::new();
+        Self::get_exported(&self.globals, &mut vals);
+        Self::get_exported(&self.locals, &mut vals);
+
+        let uni = uvars()
+            .as_ref()
+            .unwrap()
+            .get_names_ffi(true, false)
+            .from_ffi();
+        for key in uni {
+            let var = uvars()
+                .as_ref()
+                .unwrap()
+                .get_ffi(&key.to_ffi())
+                .as_ref()
+                .map(|v| v.from_ffi())
+                .expect("Variable should be present in uvars");
+            // Only insert if not already present, as uvars have lowest precedence.
+            // TODO: a longstanding bug is that an unexported local variable will not mask an exported uvar.
+            vals.entry(key).or_insert(var);
+        }
+
+        // Dorky way to add our single exported computed variable.
+        vals.insert(
+            L!("PWD").to_owned(),
+            EnvVar::new_from_name(L!("PWD"), self.perproc_data.pwd.clone()),
+        );
+
+        // Construct the export list: a list of strings of the form key=value.
+        let mut export_list: Vec<CString> = Vec::new();
+        export_list.reserve(vals.len());
+        for (key, val) in vals.into_iter() {
+            let mut str = key;
+            str.push('=');
+            str.push_utfstr(&val.as_string());
+            export_list.push(wcs2zstring(&str));
+        }
+        return Arc::new(OwningNullTerminatedArray::new(export_list));
+    }
+
+    // Exported variable array used by execv.
+    pub fn export_array(&mut self) -> Arc<OwningNullTerminatedArray> {
+        assert!(!is_forked_child());
+        if self.export_array_needs_regeneration() {
+            self.export_array = Some(self.create_export_array());
+
+            // Have to pull this into a local to satisfy the borrow checker.
+            let mut generations = std::mem::take(&mut self.export_array_generations);
```

Comment:
The C++ used clear() here. are we guaranteed that generations is empty here?

Reply:
Yikes, I missed this. Great catch. Fixed